src/NuGet.Core/NuGet.Packaging/PackageExtraction/ZipArchiveExtensions.cs

```diff
@@ -67,25 +68,65 @@ public static string SaveAsFile(this ZipArchiveEntry entry, string fileFullPath,
 
         public static void UpdateFileTimeFromEntry(this ZipArchiveEntry entry, string fileFullPath, ILogger logger)
         {
-            var attr = File.GetAttributes(fileFullPath);
+            Testable.Default.UpdateFileTimeFromEntry(entry, fileFullPath, logger);
+        }
+
+        internal class Testable
+        {
+            public static Testable Default { get; } = new Testable(EnvironmentVariableWrapper.Instance);
 
-            if (!attr.HasFlag(FileAttributes.Directory) &&
-                entry.LastWriteTime.DateTime != DateTime.MinValue && // Ignore invalid times
-                entry.LastWriteTime.UtcDateTime <= DateTime.UtcNow) // Ignore future times
+            internal Testable(IEnvironmentVariableReader environmentVariableReader)
             {
-                try
+                _updateFileTimeFromEntryMaxRetries = 9;
+                string value = environmentVariableReader.GetEnvironmentVariable("NUGET_UpdateFileTime_MaxRetries");
+                if (int.TryParse(value, out int maxRetries) && maxRetries > 0)
                 {
-                    File.SetLastWriteTimeUtc(fileFullPath, entry.LastWriteTime.Add(entry.LastWriteTime.Offset).UtcDateTime);
+                    _updateFileTimeFromEntryMaxRetries = maxRetries;
                 }
-                catch (ArgumentOutOfRangeException ex)
+            }
+
+            private readonly int _updateFileTimeFromEntryMaxRetries;
+
+            public void UpdateFileTimeFromEntry(ZipArchiveEntry entry, string fileFullPath, ILogger logger)
```

Comment:
Does this need to be `public`?

Reply:
It's on an internal class. There's no `PublicAPI.Unshipped.txt` change in this branch/PR. But ok, I can change it to internal as well.